File: 1209. Remove All Adjacent Duplicates in String II.java
class Solution {
    class Pair {
        char ch;
        int num;

        Pair(char ch, int num) {
            this.ch = ch;
            this.num = num;
        }
    }
    public String removeDuplicates(String s, int k) {
        Stack<Pair> stack = new Stack<>();

        for(int i = 0; i < s.length(); i++){
            char ch = s.charAt(i);
            if(stack.size() >= 1 && ch == stack.peek().ch){
                int count = stack.pop().num;
                stack.push(new Pair(ch, count + 1));
            }
            else{
                stack.push(new Pair(ch, 1));
            }

            while(stack.size() >= 1 && stack.peek().num >= k){
                char topCh = stack.peek().ch;
                int count = stack.pop().num;
                if(count != k){
                    stack.push(new Pair(topCh, count - k));
                }
            }
        }
        String ans = "";
        for(int i = 0; i < stack.size(); i++){
            Pair ele = stack.elementAt(i);
            for(int j = 0; j < ele.num; j++) ans += ele.ch;
        }

        return ans;
    }
}

File: 1209. Remove All Adjacent Duplicates in String II.cpp
class Solution {
    struct Pair {
        char ch;
        int num;
        Pair(char c, int n) : ch(c), num(n) {}
    };
    
public:
    string removeDuplicates(string s, int k) {
        stack<Pair> st;

        for (char ch : s) {
            if (!st.empty() && st.top().ch == ch) {
                Pair top = st.top();
                st.pop();
                st.push(Pair(ch, top.num + 1));
            } else {
                st.push(Pair(ch, 1));
            }

            // Remove group if frequency == k
            if (!st.empty() && st.top().num == k) {
                st.pop();
            }
        }

        // Build final string
        string result = "";
        stack<Pair> reversed;

        // Reverse stack to build result in correct order
        while (!st.empty()) {
            reversed.push(st.top());
            st.pop();
        }

        while (!reversed.empty()) {
            Pair p = reversed.top();
            reversed.pop();
            result.append(p.num, p.ch); // repeat char `num` times
        }

        return result;
    }
};